server/image_pull.go

```diff
@@ -27,15 +27,57 @@ func (s *Server) PullImage(ctx context.Context, req *pb.PullImageRequest) (resp
 		image = img.Image
 	}
 
+	// We use the server's pullCache to record which images are currently being
+	// pulled. This allows for avoiding pulling the same image in parallel.
+	// Hence, if a given image is currently being pulled, we queue into the
+	// pullOperation's waitgroup and wait for the pulling goroutine to unblock
+	// us and re-use its results.
+	pullOp := &pullOperation{}
+	s.pullCacheLock.Lock()
+	pr, pullInProcess := s.pullCache[image]
+	if pullInProcess {
+		pullOp = pr
+	} else {
+		s.pullCache[image] = pullOp
+		pullOp.wg.Add(1)
+	}
+	s.pullCacheLock.Unlock()
```

Comment:
I’d feel safer with a `panic`-proof locking, i.e.

```go
lock.Lock()
func() {
    defer lock.Unlock()
    …
}()
```

OTOH if we are worried about that, there should also be a `defer` to mark a `pullCache` entry as failed on failed pull, and the `defer` scopes would overlap. Oops.

Reply:
I addressed all comments but this one. We can spin off a new routine for each critical section but I feel it might be more harmful (wrt performance) in the general case. WDYT?

Reply:
Hm ... now I can't live without. I'll address this one to :^)

Reply:
Done.

Reply:
I am not one bit worried about the cost of a few stack frames, or at worst, heap allocations and atomic operations, when this is going to be blocked on downloads, decompression, and I/O in gigabytes, when trading it off for correctness.

But my initial guess was that this wasn’t going to be easy to do comprehensively anyway, so there was little point. (And, I forgot to mention, CRI-O is a top-level application, and has only one `recover` on a rare paths, so `panic`s are not _that_ much of a concern.)

For comprehensive `panic`-safety it would be necessary to also use `defer` using `defer` to make sure `pullOp.wg.Gone()` is _always_ called. And all of that makes the code very messy.

Reply:
> CRI-O is a top-level application

(I mean that libraries have to be stricter about their behavior on `panic` because they typically don’t understand and/or can’t influence the top-level application error recovery strategy.)